### PR TITLE
Bug #76407, reconcile scaled-space bottom-tabs position on Composer toggle

### PR DESCRIPTION
**Pull request by larryliang-inetsoft**
## Summary

- The Composer's `bottomTabs` checkbox toggle (`TabPropertyDialogService.setTabPropertyDialogModel`) only reconciled the master-pixel-space positions of a Tab and its children via `TabVSAssemblyInfo.repositionForBottomTabs(...)`, never the scaled/mobile-layout counterpart (`repositionForBottomTabsInScaledSpace`) -- unlike the runtime script path (`TabVSAScriptable.setBottomTabs`), which already calls both together.
- This left a child's scaled-space position stale after a plain design-time `bottomTabs` toggle, able to overlap the tab bar in mobile/responsive layout preview.
- Fix: call `TabVSAssemblyInfo.repositionForBottomTabsInScaledSpace(...)` alongside the existing master-pixel-space call, mirroring `TabVSAScriptable.setBottomTabs()`.
- `ViewsheetSandbox.executeView()`'s per-assembly bottom-tabs hook was deliberately **not** extended to also call the scaled-space method -- that method recomputes the *entire* tab bar's scaled position from all children's current extents (unlike the master-pixel `repositionChildForBottomTabs`, which only touches one child) and has no "did bottomTabs actually change" guard at that call site, so wiring it in unconditionally on every view execution risked silently overwriting a user's deliberately-set mobile-layout tab position. See `docs/teams/2026-09-02-bugs-batch1/bug-76407/03-fix.md` in the enterprise repo for the full reasoning.

## Important caveat

This fixes a **demonstrated, real gap** in the reposition mechanism (confirmed via a passing regression test showing the scaled position is left stale without this change and reconciled with it). However, it is **not confirmed** to resolve the exact symptom reported in bug #76407: the reporter's "tab" viewsheet asset does not exist anywhere in this repo and no live Composer/portal was available during this investigation, so it is unverified whether that specific viewsheet actually uses a scaled/mobile layout at all. If it doesn't, this is still a genuine improvement to the reposition mechanism, but would not be the root cause of the reported repro. Live-Composer/portal confirmation against the original repro is still needed to fully close #76407.

## Test plan

- [x] Added `testFlipToBottomTabsReconcilesScaledSpacePosition` to `TabPropertyDialogServiceTest` (existing Spring-wired, `@Tag("core")` test class) -- sets up a Tab+child with stale scaled-space positions overlapping the tab bar, flips `bottomTabs`, asserts the scaled tab position reconciles to be flush with the child.
- [x] Verified the test fails without the fix (temporarily reverted the new call, reran in isolation: `expected: <400> but was: <50>`), then restored the fix and confirmed all 7 tests in the class pass.
- [x] Ran `TabVSAScriptableTest` (12 tests) as a sanity check on the shared `TabVSAssemblyInfo` reposition methods -- unaffected, all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by larryliang-inetsoft** [comment]
## Independent review (adversarial pass)

Read the full diff (`gh pr diff 4978`), the PR description, and the affected/surrounding code directly in the worktree; ran the tests myself rather than trusting the writeup.

**Other call sites (regression check):** grepped every `bottomTabs`/`repositionForBottomTabs*`/`repositionChildForBottomTabs` reference across `core/src/main/java`. Beyond the four call sites already identified in the diagnosis/refute trail, I found several more `getBottomTabsValue()` readers not mentioned there: `CalendarPropertyDialogService`, `SelectionTreePropertyDialogService`, `SelectionListPropertyDialogService`, `ComposerObjectService` (resize/drag paths), `GroupingService`, and `VSLayoutService`. I read each — none of them toggle `bottomTabs`; they all self-adjust a single child's *master-pixel* position when something else changes (show-type, resize, drag-into-tab) while `bottomTabs` is already in a fixed state. None of them touch scaled space either (a separate, pre-existing gap, not introduced by this PR). So there is no missed call site for the specific defect this PR fixes (reconciling scaled space on a `bottomTabs` **toggle**).

**The scope-down decision (not wiring into `ViewsheetSandbox.executeView()`):** read `repositionForBottomTabsInScaledSpace()`'s full body myself (`TabVSAssemblyInfo.java:431-513`). Confirmed independently:
- It recomputes the tab bar's scaled position from **all** children's current scaled extents (`scanChildExtent`) and re-flushes **every** child — not the single-child update `repositionChildForBottomTabs` (the one `executeView()` already calls) does.
- It has no internal "did bottomTabs actually change" guard — every existing caller (`TabPropertyDialogService`, `TabVSAScriptable`) guards at the call site, not inside the method.
- It's a genuine no-op when no scaled position is set (verified the `tabScaledPos == tabInfo.getLayoutPosition(false)` reference-identity trick against `VSAssemblyInfo.getLayoutPosition(boolean)`/`setScaledPosition` directly — when `scaledPosition` is null, both calls return the same `layoutPosition` object, so the guard fires and it returns immediately). So this fix is a safe no-op for any viewsheet without a mobile/responsive layout.
- Given `executeView()`'s hook fires unconditionally on every ordinary view refresh for any `InputVSAssembly` child of a bottom-tabs `Tab`, wiring the whole-group scaled method in there (as opposed to the single-child master-pixel one already there) would risk silently overwriting a user's hand-set mobile-layout position on every render, not just on an actual toggle. I agree this is a real, non-hypothetical risk, not overcaution — scoping down to just the Composer dialog call site is the correct call.

**Test quality:** read `testFlipToBottomTabsReconcilesScaledSpacePosition` directly. It sets up a genuinely stale/overlapping scaled-space layout (tab scaled Y=50, child scaled bottom=400), flips `bottomTabs`, and asserts the tab's scaled Y becomes 400 and the child's scaled Y stays flush at 300 — this actually exercises the new call, not just its side effects. Confirmed `lenient()` is scoped correctly: `viewsheet.getID()` is read only inside the `left/top >= 0` branch of `setTabPropertyDialogModel`, and only the new test submits `left=-1, top=-1`; grepped all 7 `buildModel(...)` call sites in the file and confirmed the other 6 all submit `left/top >= 0`, so they still consume the stub normally.

**Ran the tests myself** (not just trusting the writeup): `./mvnw -o -pl core -Dtest=TabPropertyDialogServiceTest,TabVSAScriptableTest test` → `Tests run: 19, Failures: 0, Errors: 0`.

**`VSDialogService.setContainerPosition()` quirk:** confirmed genuinely unrelated and untouched by this diff. It's only invoked from `TabPropertyDialogService`'s own `if(sizePositionPaneModel.getLeft() >= 0 && getTop() >= 0)` guard (line ~192), so the new test's `left=-1, top=-1` genuinely skips it entirely — not a case of "runs but happens to produce the same result." Also independently confirmed the quirk itself is real (mixes a scaled-space `originalPosition` baseline with a submitted master-pixel `left`/`top` delta) — pre-existing, correctly left untouched per the PR's minimal-scope reasoning.

**Caveat, restated in my own words for visibility:** this fix corrects a real, demonstrated defect (Composer's design-time `bottomTabs` toggle never reconciled the scaled/mobile-layout position, unlike the runtime-script path, which already does both). It is **not confirmed** to be the root cause of the specific symptom in #76407 — the reporter's "tab" viewsheet doesn't exist anywhere in this repo (no asset, no e2e fixture), and no live browser/portal was available anywhere in this run's diagnose → refute → fix → verify pipeline to check whether that viewsheet actually uses a scaled/mobile layout. This is a real, mergeable fix for the mechanism gap it addresses; #76407 itself should stay open for live-repro confirmation.

## Verdict: **approve** — no blocker or important findings.

**Review by jshobe-inetsoft — DISMISSED** [review]
Claude's review flagged a likely-blocking issue: the new scaled-space reconciliation call in TabPropertyDialogService.java:179 appears to be immediately overwritten by the existing `dialogService.setContainerPosition(...)` call later in the same method, since `setAssemblyPosition`/`setContainerPosition` null out and replace `scaledPosition` with pixel-space coordinates whenever `left`/`top >= 0` — which is the case for the ordinary Composer "open dialog → flip bottomTabs → click OK" flow. The new test only validates the fix in isolation by submitting `left=-1, top=-1" to route around this interaction, not the realistic end-to-end flow. Please add a test with realistic (non-negative) left/top plus a scaledPosition set to confirm the fix actually holds in practice, or address the underlying position-clobbering interaction.

File: core/src/main/java/inetsoft/web/composer/vs/dialog/TabPropertyDialogService.java [hunk 1]
```diff
@@ -174,6 +174,9 @@ public Void setTabPropertyDialogModel(@ClusterProxyKey String runtimeId, String
 
          TabVSAssemblyInfo.repositionForBottomTabs(tabAssemblyInfo, vs,
                                                    tabGeneralPaneModel.getBottomTabs());
+         // scaled space: keep mobile/responsive layout preview flush too (Bug #76407)
+         TabVSAssemblyInfo.repositionForBottomTabsInScaledSpace(tabAssemblyInfo, vs,
+                                                   tabGeneralPaneModel.getBottomTabs());
```
[the remark this answers is not in the document]
Comment (larryliang-inetsoft) [reply core/src/main/java/inetsoft/web/composer/vs/dialog/TabPropertyDialogService.java:194]: Confirmed — this is a real bug, not just a theoretical gap in the test. Traced it: `setContainerPosition` runs unconditionally right after this reposition call whenever `left/top >= 0`, which is exactly the normal case since `getTabPropertyDialogModel` pre-populates the dialog from the assembly's current (scaled) position. `setLayoutPosition(Point)` unconditionally nulls `scaledPosition`, so the just-reconciled value was being immediately collapsed back to a stale master-pixel-derived one.

Fix (053dd2af2e): captured `originalScaledPos`/`hasScaledPos`/`originalScaledTop` before the reposition call, and branched the existing sync-back block to compare/sync against the *scaled* original top (instead of master) whenever a scaled position exists, so the value that survives `setContainerPosition` reflects the reconciled bottom-tabs position rather than a stale one.

Added `testFlipToBottomTabsReconcilesScaledSpacePositionRealisticFlow`, submitting realistic non-negative left/top (matching the stale scaled position) and asserting the reconciled value survives the *full* method call including `setContainerPosition`. Verified the test is load-bearing: fails with `expected: <400> but was: <50>` when the fix is disabled, passes when restored. Full suite (`TabPropertyDialogServiceTest`, 8/8) and sibling `TabVSAScriptableTest` both green with the fix in place.

File: core/src/test/java/inetsoft/web/composer/vs/dialog/TabPropertyDialogServiceTest.java [hunk 2]
```diff
@@ -277,6 +279,51 @@ void testFlipToTopTabsWithNewPosition() throws Exception {
...
+   void testFlipToBottomTabsReconcilesScaledSpacePosition() throws Exception {
+      TabVSAssemblyInfo tabInfo = (TabVSAssemblyInfo) tab.getVSAssemblyInfo();
+      // stale scaled-space layout: tab bar sits at scaled Y=50 while the child's
+      // scaled bottom edge (300 + 100 = 400) is far below it -- a large overlap.
+      tabInfo.setScaledPosition(new Point(10, 50));
+      tabInfo.setScaledSize(new Dimension(200, 30));
+      child.getVSAssemblyInfo().setScaledPosition(new Point(10, 300));
+      child.getVSAssemblyInfo().setScaledSize(new Dimension(200, 100));
+
+      TabPropertyDialogModel model = buildModel("Tab1", true, -1, -1, 200, 30);
```
[the remark this answers is not in the document]
Comment (larryliang-inetsoft) [reply core/src/test/java/inetsoft/web/composer/vs/dialog/TabPropertyDialogServiceTest.java:306]: Added the realistic-flow variant: `testFlipToBottomTabsReconcilesScaledSpacePositionRealisticFlow` submits non-negative left/top matching the current stale scaled position and a real `scaledPosition`/`scaledSize`, then asserts the reconciled scaled top survives the *entire* method call (including `setContainerPosition`), not just the reposition step. It failed with the exact predicted stale value before the corresponding fix in `TabPropertyDialogService.java`, and passes now. See the sibling thread on `TabPropertyDialogService.java:194` for the root-cause/fix explanation.

**Comment by larryliang-inetsoft** [comment]
@jshobe-inetsoft your CHANGES_REQUESTED review predates the round-2 fix (commit 053dd2af2e, pushed after your review) — it addresses exactly the setContainerPosition-collapsing-scaled-position issue the bot flagged and you cited. Both review threads are replied-to and resolved, and `build` is now passing. Could you re-review when you get a chance?

**Comment by larryliang-inetsoft** [comment]
@fredaan could you take a look and approve when you get a chance? The CHANGES_REQUESTED review predates the round-2 fix (see the reply on the review thread + the discussion above) — build now passes and both threads are resolved.

**Review by fredaan — DISMISSED** [review]

File: core/src/main/java/inetsoft/web/composer/vs/dialog/TabPropertyDialogService.java [hunk 3]
```diff
@@ -172,12 +172,39 @@ public Void setTabPropertyDialogModel(@ClusterProxyKey String runtimeId, String
...
+         // submitted back in sizePositionPaneModel is the ORIGINAL SCALED top, not the master
+         // top captured above. Capture it before reposition so the sync check below compares
+         // against the coordinate space the model's value actually came from (Bug #76407
+         // round 2 -- without this, the sync below never fires for a scaled-position tab,
+         // sizePositionPaneModel keeps the stale scaled top, and setContainerPosition's own
+         // scaled/master-collapsing behavior a few lines down overwrites the scaled position
+         // we're about to reconcile with that stale value, undoing this fix within the same
+         // request).
+         Point originalScaledPos = tabAssemblyInfo.getLayoutPosition(true);
+         boolean hasScaledPos = originalScaledPos != tabAssemblyInfo.getLayoutPosition(false);
```
[the remark this answers is not in the document]
Comment (fredaan) [reply core/src/main/java/inetsoft/web/composer/vs/dialog/TabPropertyDialogService.java]: Good catch — replaced the reference-identity check with `tabAssemblyInfo.isScaled()`. Verified it's provably identical here: `getLayoutPosition` returns the raw field reference (never a copy), so `getLayoutPosition(true) != getLayoutPosition(false)` and `scaledPosition != null` produce the same boolean in every case, not just in practice. Existing `TabPropertyDialogServiceTest` suite (8/8) still passes.

**Review by fredaan — APPROVED** [review]

**Review by fredaan — DISMISSED** [review]

**Review by Nelson-cao — APPROVED** [review]
